File: python/1206. Design Skiplist.py
import random

class Node(object):
    def __init__(self, val):
        self.val = val
        self.right = None
        self.down = None

class Skiplist(object):

    def __init__(self):
        self.head = [Node(float('-inf')) for _ in range(16)]
        self.tail = [Node(float('inf')) for _ in range(16)]

        for h, t in zip(self.head, self.tail):
            h.right = t
        for up, down in zip(self.head, self.head[1:]):
            up.down = down
        

    def search(self, target):
        """
        :type target: int
        :rtype: bool
        """
        current = self.head[0]
        while current:
            if current.val < target <= current.right.val:
                if current.right.val == target:
                    return True
                current = current.down
            else:
                current = current.right
        return False

    def add(self, num):
        """
        :type num: int
        :rtype: None
        """
        stack = []
        current = self.head[0]
        while current:
            if current.val < num <= current.right.val:
                stack.append(current)
                current = current.down
            else:
                current = current.right
        prev = None
        while stack:
            current = stack.pop()
            node = Node(num)
            node.right, current.right = current.right, node
            if prev:
                node.down = prev
            prev = node

            if random.randint(0, 1):
                break
        

    def erase(self, num):
        """
        :type num: int
        :rtype: bool
        """
        current = self.head[0]
        is_removed = False
        while current:
            if current.val < num <= current.right.val:
                if current.right.val == num:
                    current.right = current.right.right
                    is_removed = True
                current = current.down
            else:
                current = current.right
        return is_removed

        


# Your Skiplist object will be instantiated and called as such:
# obj = Skiplist()
# param_1 = obj.search(target)
# obj.add(num)
# param_3 = obj.erase(num)

File: go/src/leetcode/1206/1206. Design Skiplist.go
/**
@author ZhengHao Lou
Date    2022/07/26
*/
package main

/**
https://leetcode.cn/problems/design-skiplist/
*/
const TotalLevelNum = 30

type SkiplistNode struct {
	Val    int
	Up     *SkiplistNode
	Bottom *SkiplistNode
	Next   *SkiplistNode
}

type Skiplist struct {
	Heads [TotalLevelNum]*SkiplistNode
}

func Constructor() Skiplist {

}

func (this *Skiplist) Search(target int) bool {

}

func (this *Skiplist) Add(num int) {

}

func (this *Skiplist) Erase(num int) bool {

}

/**
 * Your Skiplist object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Search(target);
 * obj.Add(num);
 * param_3 := obj.Erase(num);
 */
